pkg/agent/bgp/gobgp/gobgp.go

```diff
@@ -0,0 +1,299 @@
+	request := &gobgpapi.DeletePeerRequest{Address: peer.Address}
+	if err := s.server.DeletePeer(ctx, request); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (s *Server) GetPeers(ctx context.Context) ([]bgp.PeerStatus, error) {
+	var peerStatuses []bgp.PeerStatus
+
+	fn := func(peer *gobgpapi.Peer) {
+		if peer == nil {
+			return
+		}
+		peerStatus := bgp.PeerStatus{}
+
+		// According to the gobgp code, all pointer fields in `peer *gobgpapi.Peer` used below should be set and non-nil
+		// when peers are listed. It's safe and harmless to keep the nil checks.
+		if transport := peer.GetTransport(); transport != nil {
+			peerStatus.Port = int32(transport.GetRemotePort())
+		}
+		if conf := peer.GetConf(); conf != nil {
+			peerStatus.Address = conf.GetNeighborAddress()
+			peerStatus.ASN = int32(conf.GetPeerAsn())
+		}
+		if ebgpMultiHop := peer.GetEbgpMultihop(); ebgpMultiHop != nil {
+			peerStatus.MultihopTTL = int32(ebgpMultiHop.GetMultihopTtl())
+		} else {
+			peerStatus.MultihopTTL = bgp.DefaultBGPMultihopTTL
+		}
+		if gracefulRestart := peer.GetGracefulRestart(); gracefulRestart != nil {
+			peerStatus.GracefulRestartTimeSeconds = int32(gracefulRestart.GetRestartTime())
+		} else {
+			peerStatus.GracefulRestartTimeSeconds = bgp.DefaultBGPGracefulRestartTimeSeconds
+		}
+		if state := peer.GetState(); state != nil {
+			peerStatus.SessionState = toSessionState(state.GetSessionState())
+			if peerStatus.SessionState == bgp.SessionEstablished {
+				if timers := peer.GetTimers(); timers != nil {
+					if timerState := timers.GetState(); timerState != nil {
+						peerStatus.UptimeSeconds = int(time.Since(timerState.GetUptime().AsTime()).Seconds())
+					}
+				}
+			}
+		}
+		peerStatuses = append(peerStatuses, peerStatus)
+	}
+
+	request := &gobgpapi.ListPeerRequest{EnableAdvertised: true}
+	err := s.server.ListPeer(ctx, request, fn)
+	if err != nil {
+		return peerStatuses, err
+	}
+	return peerStatuses, nil
+}
+
+func (s *Server) AdvertiseRoutes(ctx context.Context, routes []bgp.Route) error {
+	for i := range routes {
+		request := &gobgpapi.AddPathRequest{Path: toGoBGPPath(&routes[i])}
+		if _, err := s.server.AddPath(ctx, request); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (s *Server) WithdrawRoutes(ctx context.Context, routes []bgp.Route) error {
+	for i := range routes {
+		request := &gobgpapi.DeletePathRequest{Path: toGoBGPPath(&routes[i])}
+		if err := s.server.DeletePath(ctx, request); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (s *Server) GetRoutes(ctx context.Context, routeType bgp.RouteType, peerAddress string) ([]bgp.Route, error) {
+	var gobgpTableType gobgpapi.TableType
+	if routeType == bgp.RouteAdvertised {
+		gobgpTableType = gobgpapi.TableType_ADJ_OUT
+	} else if routeType == bgp.RouteReceived {
+		gobgpTableType = gobgpapi.TableType_ADJ_IN
+	} else {
+		return nil, fmt.Errorf("invalid route type: %v", routeType)
```

Comment:
nit: use the `%w` formatter when introducing new `fmt.Errorf` calls which wrap errors